### PR TITLE
Nested scalar support for copy if else

### LINKED ISSUE #8361: [FEA] copy_if_else support for nested scalar values
**Is your feature request related to a problem? Please describe.**
cudf::copy_if_else has support for APIs that take scalar values for both the lhs, the rhs, or both. We would like to be able to use these APIs for nested types like Structs and Arrays.

### PR DESCRIPTION
**Pull request by gerashegalov**
Uses scalar-vector-based scatter API to provide support for copy_if_else involving scalar columns. 

Other changes:
- removes some dead code
- refactoring into overloaded functions

Closes #8361, depends on #8630, #8666

### CODE REVIEW

File: cpp/src/copying/copy.cu [hunk 1]
```diff
@@ -220,11 +229,30 @@ std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
...
+
+  if constexpr (std::is_same<Left, cudf::column_view>::value &&
+                std::is_same<Right, cudf::scalar>::value) {
+    auto rhs_col = cudf::make_column_from_scalar(rhs, size, stream, mr);
+
+    return scatter_gather_based_if_else(lhs, rhs_col->view(), size, is_left, stream, mr);
+  }
+
+  // Bail out for unsupported types.
   CUDF_FAIL("Scalars of nested types are not currently supported!");
```
Comment (firestarman) [inline cpp/src/copying/copy.cu:256]: It will be better to remove this `CUDF_FAIL` or update the error message, e.g. "Unexpected types for copy".

File: cpp/src/copying/copy.cu [hunk 2]
```diff
@@ -273,7 +301,6 @@ std::unique_ptr<column> copy_if_else(Left const& lhs,
   CUDF_EXPECTS(lhs.type() == rhs.type(), "Both inputs must be of the same type");
   CUDF_EXPECTS(boolean_mask.type() == data_type(type_id::BOOL8),
                "Boolean mask column must be of type type_id::BOOL8");
-
```
Comment (firestarman) [inline cpp/src/copying/copy.cu:276]: Seems an unnecessary change ?

**Comment by firestarman** [comment]
LGTM.

Maybe in the future, we could implement this by leveraging the API `scatter` [accepting a vector of scalars](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/include/cudf/detail/scatter.hpp#L105) to avoid expanding a scalar to a column.

**Comment by revans2** [comment]
> LGTM
> 
> Maybe in the future, we could implement this by leveraging the API `scatter` [accepting a vector of scalars](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/include/cudf/detail/scatter.hpp#L105) to avoid expanding a scalar to a column.

+1 I would love to see a follow on issue filed for that.

**Comment by gerashegalov** [comment]
> > LGTM
> > Maybe in the future, we could implement this by leveraging the API `scatter` [accepting a vector of scalars](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/include/cudf/detail/scatter.hpp#L105) to avoid expanding a scalar to a column.
> 
> +1 I would love to see a follow on issue filed for that.

I agree, I reflected in the PR description that this is a naive implementation. Depending on whether this iteration of the PR is approved, I'll file a follow-up issue on improving it. 

@nvdbaranec , @mythrocks please take a look,

File: cpp/src/copying/copy.cu [hunk 3]
```diff
@@ -220,18 +229,31 @@ std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
...
-  //   1. Struct scalars are not yet available.
-  //   2. List scalars do not yet support explosion to a full column.
-  CUDF_FAIL("Scalars of nested types are not currently supported!");
-  (void)lhs;
-  (void)rhs;
-  (void)size;
-  (void)is_left;
-  (void)stream;
-  (void)mr;
+  if constexpr (std::is_same<Left, cudf::scalar>::value &&
```
Comment (mythrocks) [inline cpp/src/copying/copy.cu:232]: Since there are now 4 paths generated here, in the interest of brevity, I wonder if we can reduce the code repetition, via a helper template.

I'll experiment, and offer a suggestion shortly.

Comment (mythrocks) [reply cpp/src/copying/copy.cu:232]: The other consideration is whether the "paths not taken" are actually eliminated by `nvcc`. Last I checked, both `nvcc-11` and `gcc-9` were good about this. But I'd feel better if we confirmed.

Comment (mythrocks) [reply cpp/src/copying/copy.cu:232]: The more I look at this, the more I think we should do this with iterators at some point, rather than inflate the struct scalar to a column. :/

Comment (mythrocks) [reply cpp/src/copying/copy.cu:232]: > Since there are now 4 paths generated here, in the interest of brevity, I wonder if we can reduce the code repetition, via a helper template.

Sorry for the delay. Here is the helper template I was suggesting:
```c++
template <typename T>
auto get_column_and_view(T const& t, 
                         cudf::size_type,
                         rmm::cuda_stream_view, 
                         rmm::mr::device_memory_resource*)
{
  return std::make_tuple(t, t);
}

template <>
auto get_column_and_view<cudf::scalar>(cudf::scalar const& scalar, 
                                   cudf::size_type size,
                                   rmm::cuda_stream_view stream, 
                                   rmm::mr::device_memory_resource* mr)
{
  auto column = cudf::make_column_from_scalar(scalar, size, stream, mr);
  return std::make_tuple(std::move(column), column->view());
}
```
This transparently returns a column and a view to it, regardless of whether the input is a `column_view` or a `scalar`.

`scatter_gather_based_if_else()` can now be reduced to:
```c++
template <typename Left, typename Right, typename Filter>
std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
                                                     Right const& rhs,
                                                     size_type size,
                                                     Filter is_left,
                                                     rmm::cuda_stream_view stream,
                                                     rmm::mr::device_memory_resource* mr)
{
  auto const [lhs_col, lhs_view] = get_column_and_view(lhs, size, stream, mr);
  auto const [rhs_col, rhs_view] = get_column_and_view(rhs, size, stream, mr);

  return scatter_gather_column_based_if_else(lhs_view, rhs_view, size, is_left, stream, mr);
}
```
`scatter_gather_column_based_if_else()` simply has the `column_view` based scatter-based `if-else`, without any `if constexpr` for anything.

Does this look agreeable?

File: cpp/src/copying/copy.cu [hunk 4]
```diff
@@ -220,18 +229,31 @@ std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
...
+                std::is_same<Right, cudf::scalar>::value) {
+    auto lhs_col = cudf::make_column_from_scalar(lhs, size, stream, mr);
+    auto rhs_col = cudf::make_column_from_scalar(rhs, size, stream, mr);
+
+    return scatter_gather_based_if_else(
+      lhs_col->view(), rhs_col->view(), size, is_left, stream, mr);
+  }
+
+  if constexpr (std::is_same<Left, cudf::scalar>::value &&
```
Comment (mythrocks) [inline cpp/src/copying/copy.cu:241]: Since we're adding `scalar` support, you'll want to change the function's documentation to remove the `Scalars are not supported.` bit. :]

File: cpp/src/copying/copy.cu (hunk 3 above)
Comment (mythrocks) [reply cpp/src/copying/copy.cu:232]: Please forgive my pointing out that `get_column_and_view()` has a couple of gotchas:

First, the function can't just return the column view. It must return the column, to keep the view alive.

Secondly, we can't `return std::make_tuple(std::move(column), column->view())` in one go. It's not reliable, because we don't want `column->view()` called *after* `std::move(column)`.
Switching the expressions around doesn't guarantee anything. The order of evaluation of those sub-expressions isn't guaranteed.

Comment (nvdbaranec) [reply cpp/src/copying/copy.cu:232]: > 
> 
> The more I look at this, the more I think we should do this with iterators at some point, rather than inflate the struct scalar to a column. :/

Yah, Gera and I talked about this offline, but the fundamental issue is that gather itself doesn't take iterators as column inputs (it does for the map though).  I think that's a pretty hard change, but it would be pretty nice. Doing so would also allow things like "multi-gather" where you can stitch together multiple columns together at once.

**Comment by gerashegalov** [comment]
> > > LGTM
> > > Maybe in the future, we could implement this by leveraging the API `scatter` [accepting a vector of scalars](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/include/cudf/detail/scatter.hpp#L105) to avoid expanding a scalar to a column.
> > 
> > 
> > +1 I would love to see a follow on issue filed for that.
> 
> I agree, I reflected in the PR description that this is a naive implementation. Depending on whether this iteration of the PR is approved, I'll file a follow-up issue on improving it.
> 
> @nvdbaranec , @mythrocks please take a look, too

prototyped this in https://github.com/gerashegalov/cudf/pull/1 but it will need more work for structs. So I think it should be definitely a follow-up PR

**Comment by gerashegalov** [comment]
> > LGTM
> > Maybe in the future, we could implement this by leveraging the API `scatter` [accepting a vector of scalars](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/include/cudf/detail/scatter.hpp#L105) to avoid expanding a scalar to a column.
> 
> +1 I would love to see a follow on issue filed for that.

#8630 went in. So I folded  gerashegalov#1 into this PR that uses scalar-vector-based API

**Comment by firestarman** [comment]
rerun tests

File: cpp/src/copying/copy.cu [hunk 5]
```diff
@@ -183,55 +159,89 @@ class logical_not {
...
+                                                     rmm::mr::device_memory_resource* mr)
+{
+  auto scatter_map = rmm::device_uvector<size_type>{static_cast<std::size_t>(size), stream};
+  auto const scatter_map_end = thrust::copy_if(rmm::exec_policy(stream),
+                                               thrust::make_counting_iterator(size_type{0}),
+                                               thrust::make_counting_iterator(size_type{size}),
+                                               scatter_map.begin(),
+                                               is_left);
+
+  auto const scatter_map_size  = scatter_map_end - scatter_map.begin();
```
Comment (mythrocks) [inline cpp/src/copying/copy.cu:212]: Nitpick: I have been corrected in the past to use `std::distance()` instead of explicit subtraction, for readability:
```suggestion
  auto const scatter_map_size  = std::distance(scatter_map.begin(), scatter_map_end);
```

File: cpp/src/copying/copy.cu [hunk 6]
```diff
@@ -183,55 +159,89 @@ class logical_not {
...
+                                                   scatter_map.begin()};
+
+  auto result = cudf::scatter(
+    scatter_source, scatter_map_column_view, table_view{std::vector<column_view>{rhs}}, false, mr);
+
+  return std::move(result->release()[0]);
+}
+
+template <typename Filter>
+std::unique_ptr<column> scatter_gather_based_if_else(cudf::column_view const& lhs,
```
Comment (mythrocks) [inline cpp/src/copying/copy.cu:225]: It breaks my heart that these overloads can't be collapsed into a single function. I'll play with this a little bit and see if I can't come up with something.

Comment (mythrocks) [reply cpp/src/copying/copy.cu:225]: I think I see the problem. The `detail::scatter()` APIs are not symmetric across the `table_view` and `vector<reference<scalar>>` overloads. :/
If we had an overload of `detail::scatter(vector<reference<scalar>>, ...)` that took `[scatter_map_begin, scatter_map_end)` as arguments instead of a `column_view`, we stood a chance of collapsing these 4 overloads into one or two. 

I think it can be done, based on the definitions I see in [scatter.cu](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/src/copying/scatter.cu#L308).

That said, I'm not opposed to leaving this the way it is, if it looks like too much change.

Comment (gerashegalov) [reply cpp/src/copying/copy.cu:225]: My preference is to do this refactoring as a separate PR

File: cpp/src/copying/copy.cu [hunk 7]
```diff
@@ -183,55 +159,89 @@ class logical_not {
 /**
  * @brief Implementation of copy_if_else() with gather()/scatter().
  *
- * Currently supports only nested-type column_views. Scalars are not supported.
+ * Handles nested-typed columns and scalars
```
Comment (nvdbaranec) [inline cpp/src/copying/copy.cu:162]: I think this comment is wrong now.  Function appears to be column_view only.

File: cpp/src/copying/copy.cu [hunk 8]
```diff
@@ -183,55 +159,89 @@ class logical_not {
...
+  auto result = cudf::detail::scatter(
+    table_view{std::vector<column_view>{scatter_src_lhs->get_column(0).view()}},
+    scatter_map.begin(),
+    scatter_map_end,
+    table_view{std::vector<column_view>{rhs}},
+    false,
+    stream,
+    mr);
+
+  return std::move(result->release()[0]);
```
Comment (nvdbaranec) [inline cpp/src/copying/copy.cu:194]:
```suggestion
  return result->release()[0];
```

File: cpp/src/copying/copy.cu [hunk 9]
```diff
@@ -183,55 +159,89 @@ class logical_not {
...
+  auto const scatter_map_size  = std::distance(scatter_map.begin(), scatter_map_end);
+  auto scatter_source          = std::vector<std::reference_wrapper<const scalar>>{std::ref(lhs)};
+  auto scatter_map_column_view = cudf::column_view{cudf::data_type{cudf::type_id::INT32},
+                                                   static_cast<cudf::size_type>(scatter_map_size),
+                                                   scatter_map.begin()};
+
+  auto result = cudf::scatter(
+    scatter_source, scatter_map_column_view, table_view{std::vector<column_view>{rhs}}, false, mr);
+
+  return std::move(result->release()[0]);
```
Comment (nvdbaranec) [inline cpp/src/copying/copy.cu:221]:
```suggestion
  return result->release()[0];
```

File: cpp/src/copying/copy.cu [hunk 10]
```diff
@@ -183,55 +159,89 @@ class logical_not {
...
  * @brief Implementation of copy_if_else() with gather()/scatter().
  *
- * Currently supports only nested-type column_views. Scalars are not supported.
+ * Handles nested-typed columns and scalars
  */
-template <typename Left, typename Right, typename Filter>
-std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
-                                                     Right const& rhs,
+template <typename Filter>
+std::unique_ptr<column> scatter_gather_based_if_else(cudf::column_view const& lhs,
```
Comment (nvdbaranec) [inline cpp/src/copying/copy.cu:165]: I think a doxygen block would be a good idea for this function.  It's not clear what "is_left" means at a glance.

**Review by nvdbaranec — CHANGES_REQUESTED** [review]

File: cpp/src/copying/copy.cu (hunk 9 above)
Comment (gerashegalov) [reply cpp/src/copying/copy.cu:221]: this results in compile errors

Comment (nvdbaranec) [reply cpp/src/copying/copy.cu:233]: Right.  Misinterpreted what was being released here.  Sorry.

File: cpp/src/copying/copy.cu [hunk 11]
```diff
@@ -159,7 +159,19 @@ class logical_not {
...
+ * Uses `rhs` as the destination for scatter. First gathers indices of rows to copy from lhs.
+ *
+ * @tparam Filter Bool iterator producing `true` for indices of output rows to copy from `lhs` and
+ * `false` for indices of output rows to copy from `rhs`
+ * @param lhs Left-hand side input column view
+ * @param rhs Right-hand side input column view
+ * @param size The size of the output column, inputs rows are iterated from 0 to `size - 1`
+ * @param is_left Predicate for picking rows from `lhs` on `true` or `rhs` on `false`
+ * @param stream The stream on which to perform the allocation
+ * @param mr The resource used to allocate the device storage
```
Comment (nvdbaranec) [inline cpp/src/copying/copy.cu:174]: Needs `@returns `

**Review by nvdbaranec — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]
:+1: LGTM

**Comment by gerashegalov** [comment]
rerun tests

**Comment by gerashegalov** [comment]
@gpucibot merge
